src/Database/LSMTree/Internal/RunBuilder.hs

```diff
@@ -177,44 +174,41 @@ close fs RunBuilder {..} = do
 
 writeRawPage :: HasFS IO h -> RunBuilder (FS.Handle h) -> RawPage -> IO ()
 writeRawPage fs RunBuilder {..} =
-    --TODO: avoid copying via builder and write the pinned RawPage directly
     writeToHandle fs (forRunKOps runBuilderHandles)
-  . RB.builder
+  . BSL.fromStrict
+  . RB.unsafePinnedToByteString -- 'RawPage' is guaranteed to be pinned
   . RawPage.rawPageRawBytes
 
-writeRawOverflowPage :: HasFS IO h -> RunBuilder (FS.Handle h) -> RawOverflowPage -> IO ()
-writeRawOverflowPage fs RunBuilder {..} =
-    --TODO: avoid copying via builder and write the pinned RawPage directly
+writeRawOverflowPages :: HasFS IO h -> RunBuilder (FS.Handle h) -> [RawOverflowPage] -> IO ()
+writeRawOverflowPages fs RunBuilder {..} =
     writeToHandle fs (forRunKOps runBuilderHandles)
-  . RB.builder
-  . RawOverflowPage.rawOverflowPageRawBytes
+  . BSL.fromChunks
+  . map (RB.unsafePinnedToByteString . RawOverflowPage.rawOverflowPageRawBytes)
+    -- 'RawOverflowPage' is guaranteed to use a pinned byte array
```

Comment:
It'd be a tad neater to make a `rawOverflowPageToByteString` util and put it in the `RawOverflowPage` module, since that's where we have the guarantee about the raw overflow pages being pinned, rather than relying on that special knowledge here at the call site.